app/cluster/state/etcd.go

```diff
@@ -0,0 +1,279 @@
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, true, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, true, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		ReleaseName:          releaseName,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (manager *ETCDManager) prepareMode(raw []byte) servermode.ModeRequest {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ModeError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := servermode.Mode(req.Mode)
+	if !mode.Valid() {
+		return servermode.ModeError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewModeRequest(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return err
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ModeRequest {
+	ch := make(chan servermode.ModeRequest, 1)
+	ch <- servermode.ModeError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ModeRequest {
+	if err := manager.init(); err != nil {
+		return errChModeRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(modeRequestKeyPattern, manager.Config.ReleaseName, manager.Config.ServerIndex)
+
+	resultChan := make(chan servermode.ModeRequest, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChModeRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChModeRequest(errors.New("no workspace found"))
+	}
+
+	resultChan <- manager.prepareMode(resp.Kvs[0].Value)
+
+	etcdWatchChan := manager.Client.Watch(ctx, modeRequestKey, clientv3.WithRev(resp.Header.Revision+1))
+	go func() {
```

Comment:
Shouldn't you propagate the context in these two for loops? The same goes for the `WorkspaceIDs` method.

Reply:
I am not sure I understand what you mean by that.

Context cancellation will cause `.Watch` to close the channel, thus the loops will stop and the go routine will return.

Reply:
Right, so it is coupled to `.Watch` which also happens to be a library we don't have control over since it's a 3rd party.

I meant that instead of:

```go
for watchResp := range etcdWatchChan {
```

One should do:

```go
for {
    select {
    case <-ctx.Done():
        return ctx.Err()
    case watchResp := <-etcdWatchCan:
        // do your magic
```

Reply:
I disagree, the producer is responsible for closing the channel. In this case, the producer is etcd client.
They are very clear in their documentation of how this should be used:
> 	// Watch watches on a key or prefix. The watched events will be returned
	// through the returned channel. If revisions waiting to be sent over the
	// watch are compacted, then the watch will be canceled by the server, the
	// client will post a compacted error watch response, and the channel will close.
	// If the requested revision is 0 or unspecified, the returned channel will
	// return watch events that happen after the server receives the watch request.
	// If the context "ctx" is canceled or timed out, returned "WatchChan" is closed,
	// and "WatchResponse" from this closed channel has zero events and nil "Err()".
	// The context "ctx" MUST be canceled, as soon as watcher is no longer being used,
	// to release the associated resources.
	//
	// If the context is "context.Background/TODO", returned "WatchChan" will
	// not be closed and block until event is triggered, except when server
	// returns a non-recoverable error (e.g. ErrCompacted).
	// For example, when context passed with "WithRequireLeader" and the
	// connected server has no leader (e.g. due to network partition),
	// error "etcdserver: no leader" (ErrNoLeader) will be returned,
	// and then "WatchChan" is closed with non-nil "Err()".
	// In order to prevent a watch stream being stuck in a partitioned node,
	// make sure to wrap context with "WithRequireLeader".

Reply:
That's good enough :+1:

Reply:
I think there is some miss understanding. Context cancellation will cause `.Watch` to close the channel it returns:
> If the context "ctx" is canceled or timed out, returned "WatchChan" is closed

In our `.ServerMode`  / `.WorkspaceIDs` implementation when `etcdWatchChan` is closed, the for loops are exited and we close the `resultChan`. 

> Pushing to that channel while the context has been canceled can still happen

Yes, for me this is not necessarily bad. I understand that it can delay the shutdown, for me, this is like an inflight request (the scheduler request something, so we are going to do it and awk and then shutdown). If we see that this creates problems then we can ignore it.